File: android/src/main/java/com/rnlibrary/wechat/RNLWeChatReq.java
package com.rnlibrary.wechat;

import android.content.Context;

import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.tencent.mm.opensdk.modelbiz.AddCardToWXCardPackage;
import com.tencent.mm.opensdk.modelbiz.ChooseCardFromWXCardPackage;
import com.tencent.mm.opensdk.modelbiz.JumpToBizProfile;
import com.tencent.mm.opensdk.modelbiz.JumpToBizTempSession;
import com.tencent.mm.opensdk.modelbiz.JumpToBizWebview;
import com.tencent.mm.opensdk.modelbiz.OpenRankList;
import com.tencent.mm.opensdk.modelbiz.OpenWebview;
import com.tencent.mm.opensdk.modelbiz.SubscribeMessage;
import com.tencent.mm.opensdk.modelbiz.SubscribeMiniProgramMsg;
import com.tencent.mm.opensdk.modelbiz.WXInvoiceAuthInsert;
import com.tencent.mm.opensdk.modelbiz.WXLaunchMiniProgram;
import com.tencent.mm.opensdk.modelbiz.WXNontaxPay;
import com.tencent.mm.opensdk.modelbiz.WXOpenBusinessView;
import com.tencent.mm.opensdk.modelbiz.WXOpenBusinessWebview;
import com.tencent.mm.opensdk.modelbiz.WXPayInsurance;
import com.tencent.mm.opensdk.modelmsg.SendAuth;
import com.tencent.mm.opensdk.modelmsg.SendMessageToWX;
import com.tencent.mm.opensdk.modelpay.JumpToOfflinePay;
import com.tencent.mm.opensdk.modelpay.PayReq;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class RNLWeChatReq {
    static SendAuth.Req getSendAuthReq(ReadableMap option) {
        SendAuth.Req req = new SendAuth.Req();
        req.scope = option.getString("scope");
        req.state = option.getString("state");
        return req;
    }

    static PayReq getPayReq(ReadableMap option, String appID) {
        PayReq req = new PayReq();
        req.appId = appID;
        req.partnerId = option.getString("partnerId");
        req.prepayId = option.getString("prepayId");
        req.nonceStr = option.getString("nonce");
        req.timeStamp = String.valueOf(option.getInt("timestamp"));
        req.packageValue = option.getString("package");
        req.sign = option.getString("sign");
        return req;
    }

    static JumpToOfflinePay.Req getJumpToOfflinePayReq() {
        return new JumpToOfflinePay.Req();
    }

    static WXNontaxPay.Req getWXNontaxPayReq(ReadableMap option) {
        WXNontaxPay.Req req = new WXNontaxPay.Req();
        req.url = option.getString("url");
        return req;
    }

    static WXPayInsurance.Req getWXPayInsuranceReq(ReadableMap option) {
        WXPayInsurance.Req req = new WXPayInsurance.Req();
        req.url = option.getString("url");
        return req;
    }

    static JumpToBizTempSession.Req getJumpToBizTempSessionReq(ReadableMap option) {
        JumpToBizTempSession.Req req = new JumpToBizTempSession.Req();
        req.toUserName = option.getString("username");
        req.sessionFrom = option.getString("session");
        return req;
    }

    static OpenRankList.Req getOpenRankListReq() {
        return new OpenRankList.Req();
    }

    static OpenWebview.Req getOpenWebviewReq(ReadableMap option) {
        OpenWebview.Req req = new OpenWebview.Req();
        req.url = option.getString("url");
        return req;
    }

    static WXOpenBusinessView.Req getWXOpenBusinessViewReq(ReadableMap option) {
        WXOpenBusinessView.Req req = new WXOpenBusinessView.Req();
        req.businessType = option.getString("type");
        req.query = option.getString("type");
        req.extInfo = option.getString("ext");
        return req;
    }

    static WXOpenBusinessWebview.Req getWXOpenBusinessWebviewReq(ReadableMap option) {
        WXOpenBusinessWebview.Req req = new WXOpenBusinessWebview.Req();
        req.businessType = option.getInt("type");
        HashMap<String, String> query = new HashMap<>();
        for (Map.Entry<String, Object> item : option.getMap("query").toHashMap().entrySet()) {
            query.put(item.getKey(), (String) item.getValue());
        }
        req.queryInfo = query;
        return req;
    }

    static JumpToBizProfile.Req getJumpToBizProfileReq(ReadableMap option) {
        JumpToBizProfile.Req req = new JumpToBizProfile.Req();
        req.profileType = option.getInt("type");
        req.toUserName = option.getString("username");
        req.extMsg = option.getString("ext");
        return req;
    }

    static JumpToBizWebview.Req getJumpToBizWebviewReq(ReadableMap option) {
        JumpToBizWebview.Req req = new JumpToBizWebview.Req();
        req.webType = option.getInt("type");
        req.toUserName = option.getString("username");
        req.extMsg = option.getString("ext");
        return req;
    }

    static AddCardToWXCardPackage.Req getAddCardToWXCardPackageReq(ReadableMap option) {
        AddCardToWXCardPackage.Req req = new AddCardToWXCardPackage.Req();
        ReadableArray cardsData = option.getArray("cards");
        if (cardsData != null) {
            List<AddCardToWXCardPackage.WXCardItem> cards = new ArrayList<>();
            for (int i = 0; i < cardsData.size(); i++) {
                ReadableMap cardItem = cardsData.getMap(i);
                AddCardToWXCardPackage.WXCardItem item = new AddCardToWXCardPackage.WXCardItem();
                item.cardId = cardItem.getString("id");
                item.cardExtMsg = cardItem.getString("ext");
                cards.add(item);
            }
            req.cardArrary = cards;
        }
        return req;
    }

    static ChooseCardFromWXCardPackage.Req getChooseCardFromWXCardPackageReq(ReadableMap option) {
        ChooseCardFromWXCardPackage.Req req = new ChooseCardFromWXCardPackage.Req();
        req.appId = option.getString("appID");
        req.locationId = String.valueOf(option.getInt("shopID"));
        req.canMultiSelect = String.valueOf(option.getBoolean("multiSelect"));
        req.cardType = option.getString("cardType");
        req.cardId = option.getString("cardTpID");
        req.signType = option.getString("signType");
        req.cardSign = option.getString("cardSign");
        req.timeStamp = String.valueOf(option.getInt("timestamp"));
        req.nonceStr = option.getString("nonce");
        return req;
    }

    static WXInvoiceAuthInsert.Req getWXInvoiceAuthInsertReq(ReadableMap option) {
        WXInvoiceAuthInsert.Req req = new WXInvoiceAuthInsert.Req();
        req.url = option.getString("url");
        return req;
    }

    static WXLaunchMiniProgram.Req getWXLaunchMiniProgramReq(ReadableMap option) {
        WXLaunchMiniProgram.Req req = new WXLaunchMiniProgram.Req();
        req.userName = option.getString("username");
        req.path = option.getString("path");
        req.miniprogramType = option.getInt("type");
        req.extData = option.getString("ext");
        return req;
    }

    static SubscribeMiniProgramMsg.Req getSubscribeMiniProgramMsgReq(ReadableMap option) {
        SubscribeMiniProgramMsg.Req req = new SubscribeMiniProgramMsg.Req();
        req.miniProgramAppId = option.getString("appID");
        return req;
    }

    static SendMessageToWX.Req getSendMessageToWXReq(Context context, ReadableMap option) throws Exception {
        SendMessageToWX.Req req = new SendMessageToWX.Req();
        req.scene = option.getInt("scene");
        req.message = RNLWeChatAPIObjectConvert.mediaMessageToWXO(context, option);
        req.userOpenId = option.getString("userOpenID");
        return req;
    }

    static SubscribeMessage.Req getSubscribeMessageReq(ReadableMap option) {
        SubscribeMessage.Req req = new SubscribeMessage.Req();
        req.scene = option.getInt("scene");
        req.templateID = option.getString("templateID");
        req.reserved = option.getString("reserved");
        return req;
    }
}

File: ios/RNLWeChatReq.m
#import <React/RCTConvert.h>

#import "RNLWeChatReq.h"
#import "RNLWeChatAPIObjectConvert.h"

@implementation RNLWeChatReq

+ (SendAuthReq *)getSendAuthReqWithOption:(NSDictionary *)option
{
    SendAuthReq *req = [SendAuthReq new];
    req.scope = [RCTConvert NSString:option[@"scope"]];
    req.state = [RCTConvert NSString:option[@"state"]];
    return req;
}

+ (PayReq *)getPayReqWithOption:(NSDictionary *)option
{
    PayReq *req = [PayReq new];
    req.partnerId = [RCTConvert NSString:option[@"partnerId"]];
    req.prepayId = [RCTConvert NSString:option[@"prepayId"]];
    req.nonceStr = [RCTConvert NSString:option[@"nonce"]];
    req.timeStamp = (UInt32)[RCTConvert NSUInteger:option[@"timestamp"]];
    req.package = [RCTConvert NSString:option[@"package"]];
    req.sign = [RCTConvert NSString:option[@"sign"]];
    return req;
}

+ (WXOfflinePayReq *)getWXOfflinePayReqWithOption:(NSDictionary *)option;
{
    WXOfflinePayReq *req = [WXOfflinePayReq new];
    return req;
}

+ (WXNontaxPayReq *)getWXNontaxPayReqWithOption:(NSDictionary *)option;
{
    WXNontaxPayReq *req = [WXNontaxPayReq new];
    req.urlString = [RCTConvert NSString:option[@"url"]];
    return req;
}

+ (WXPayInsuranceReq *)getWXPayInsuranceReqWithOption:(NSDictionary *)option;
{
    WXPayInsuranceReq *req = [WXPayInsuranceReq new];
    req.urlString = [RCTConvert NSString:option[@"url"]];
    return req;
}

+ (OpenTempSessionReq *)getOpenTempSessionReqWithOption:(NSDictionary *)option;
{
    OpenTempSessionReq *req = [OpenTempSessionReq new];
    req.username = [RCTConvert NSString:option[@"username"]];
    req.sessionFrom = [RCTConvert NSString:option[@"session"]];
    return req;
}

+ (OpenRankListReq *)getOpenRankListReqWithOption:(NSDictionary *)option;
{
    OpenRankListReq *req = [OpenRankListReq new];
    return req;
}

+ (OpenWebviewReq *)getOpenWebViewReqWithOption:(NSDictionary *)option;
{
    OpenWebviewReq *req = [OpenWebviewReq new];
    req.url = [RCTConvert NSString:option[@"url"]];
    return req;
}

+ (WXOpenBusinessViewReq *)getWXOpenBusinessViewReqWithOption:(NSDictionary *)option;
{
    WXOpenBusinessViewReq *req = [WXOpenBusinessViewReq new];
    req.businessType = [RCTConvert NSString:option[@"type"]];
    req.query = [RCTConvert NSString:option[@"query"]];
    req.extInfo = [RCTConvert NSString:option[@"ext"]];
    return req;
}

+ (WXOpenBusinessWebViewReq *)getWXOpenBusinessWebViewReqWithOption:(NSDictionary *)option;
{
    WXOpenBusinessWebViewReq *req = [WXOpenBusinessWebViewReq new];
    req.businessType = (UInt32)[RCTConvert NSUInteger:option[@"type"]];
    req.queryInfoDic = [RCTConvert NSDictionary:option[@"query"]];
    return req;
}

+ (JumpToBizProfileReq *)getJumpToBizProfileReqWithOption:(NSDictionary *)option
{
    JumpToBizProfileReq *req = [JumpToBizProfileReq new];
    req.profileType = [RCTConvert int:option[@"type"]];
    req.username = [RCTConvert NSString:option[@"username"]];
    req.extMsg = [RCTConvert NSString:option[@"ext"]];
    return req;
}

+ (JumpToBizWebviewReq *)getJumpToBizWebviewReqWithOption:(NSDictionary *)option
{
    JumpToBizWebviewReq *req = [JumpToBizWebviewReq new];
    req.webType = [RCTConvert int:option[@"type"]];
    req.tousrname = [RCTConvert NSString:option[@"username"]];
    req.extMsg = [RCTConvert NSString:option[@"ext"]];
    return req;
}

+ (AddCardToWXCardPackageReq *)getAddCardToWXCardPackageReqWithOption:(NSDictionary *)option;
{
    AddCardToWXCardPackageReq *req = [AddCardToWXCardPackageReq new];
    NSMutableArray *cards = [NSMutableArray new];
    for (NSDictionary *cardData in [RCTConvert NSArray:option[@"cards"]]) {
        WXCardItem *card = [WXCardItem new];
        card.cardId = [RCTConvert NSString:cardData[@"id"]];
        card.extMsg = [RCTConvert NSString:cardData[@"ext"]];
        [cards addObject:card];
    }
    req.cardAry = cards;
    return req;
}

+ (WXChooseCardReq *)getWXChooseCardReqWithOption:(NSDictionary *)option;
{
    WXChooseCardReq *req = [WXChooseCardReq new];
    req.appID = [RCTConvert NSString:option[@"appID"]];
    req.shopID = (UInt32)[RCTConvert NSUInteger:option[@"shopID"]];
    req.canMultiSelect = (UInt32)[RCTConvert BOOL:option[@"multiSelect"]];
    req.cardType = [RCTConvert NSString:option[@"cardType"]];
    req.cardTpID = [RCTConvert NSString:option[@"cardTpID"]];
    req.signType = [RCTConvert NSString:option[@"signType"]];
    req.cardSign = [RCTConvert NSString:option[@"cardSign"]];
    req.timeStamp = (UInt32)[RCTConvert NSUInteger:option[@"timestamp"]];
    req.nonceStr = [RCTConvert NSString:option[@"nonce"]];
    return req;
}

+ (WXChooseInvoiceReq *)getWXChooseInvoiceReqWithOption:(NSDictionary *)option;
{
    WXChooseInvoiceReq *req = [WXChooseInvoiceReq new];
    req.appID = [RCTConvert NSString:option[@"appID"]];
    req.shopID = (UInt32)[RCTConvert NSUInteger:option[@"shopID"]];
    req.signType = [RCTConvert NSString:option[@"signType"]];
    req.cardSign = [RCTConvert NSString:option[@"cardSign"]];
    req.timeStamp = (UInt32)[RCTConvert NSUInteger:option[@"timestamp"]];
    req.nonceStr = [RCTConvert NSString:option[@"nonce"]];
    return req;
}

+ (WXInvoiceAuthInsertReq *)getWXInvoiceAuthInsertReqWithOption:(NSDictionary *)option;
{
    WXInvoiceAuthInsertReq *req = [WXInvoiceAuthInsertReq new];
    req.urlString = [RCTConvert NSString:option[@"url"]];
    return req;
}

+ (WXLaunchMiniProgramReq *)getWXLaunchMiniProgramReqWithOption:(NSDictionary *)option;
{
    WXLaunchMiniProgramReq *req = [WXLaunchMiniProgramReq new];
    req.userName = [RCTConvert NSString:option[@"username"]];
    req.path = [RCTConvert NSString:option[@"path"]];
    req.miniProgramType = [RCTConvert NSUInteger:option[@"type"]];
    req.extMsg = [RCTConvert NSString:option[@"ext"]];
    return req;
}

+ (WXSubscribeMiniProgramMsgReq *)getWXSubscribeMiniProgramMsgReqWithOption:(NSDictionary *)option;
{
    WXSubscribeMiniProgramMsgReq *req = [WXSubscribeMiniProgramMsgReq new];
    req.miniProgramAppid = [RCTConvert NSString:option[@"appID"]];
    return req;
}

+ (SendMessageToWXReq *)getSendMessageToWXReqWithOption:(NSDictionary *)option;
{
    SendMessageToWXReq *req = [SendMessageToWXReq new];
    req.scene = [RCTConvert int:option[@"scene"]];
    if ([RCTConvert NSInteger:option[@"messageType"]] == RNLWeChatTextMessage) {
        req.bText = YES;
        req.text = [RCTConvert NSString:option[@"content"]];
    } else {
        req.bText = NO;
    }
    req.message = [RNLWeChatAPIObjectConvert MediaMessageToWXO:option];
    req.toUserOpenId = [RCTConvert NSString:option[@"userOpenID"]];
    return req;
}

+ (WXSubscribeMsgReq *)getWXSubscribeMsgReqWithOption:(NSDictionary *)option;
{
    WXSubscribeMsgReq *req = [WXSubscribeMsgReq new];
    req.scene = (UInt32)[RCTConvert NSUInteger:option[@"scene"]];
    req.templateId = [RCTConvert NSString:option[@"templateID"]];
    req.reserved = [RCTConvert NSString:option[@"reserved"]];
    return req;
}

@end
